File: sdk_python/evidence/chain.py
"""Evidence-stream linking, fork detection, and key continuity (ES-006..008/024)."""

from __future__ import annotations

import base64
import binascii
import re
from collections.abc import Iterable, Mapping
from dataclasses import dataclass
from typing import Protocol

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

from sdk_python.evidence.canonical import canonical_digest, canonicalize
from sdk_python.evidence.origin import primary_signer_namespace
from sdk_python.evidence.schema import RecordEnvelope
from sdk_python.evidence.signing import (
    ALGORITHM,
    CONTINUITY_FIELDS,
    SignatureError,
    UnsupportedAlgorithmError,
    verify_record_signature,
)


class ChainVerificationError(ValueError):
    """A break with enough context for the attestation layer to stop safely."""

    def __init__(
        self,
        message: str,
        *,
        stream_id: str | None = None,
        break_sequence: int | None = None,
        valid_through_sequence: int | None = None,
        fatal: bool = False,
        coverage_gap_body: Mapping[str, object] | None = None,
    ) -> None:
        super().__init__(message)
        self.stream_id = stream_id
        self.break_sequence = break_sequence
        self.valid_through_sequence = valid_through_sequence
        self.attestation_permitted = not fatal
        self.coverage_gap_body = dict(coverage_gap_body or {})


class StreamForkError(ChainVerificationError):
    """Two different records claim one stream position; no prefix is trusted."""


class SequenceGapError(ChainVerificationError):
    """A sequence number is missing from the submitted stream."""


class DigestLinkError(ChainVerificationError):
    """A record does not commit the canonical digest of its predecessor."""


class KeyContinuityError(ChainVerificationError):
    """A signing-key change lacks a valid proof from the predecessor key."""


class KeyNamespaceError(ChainVerificationError):
    """A cryptographically valid signer belongs to a forbidden key namespace."""


class RegisteredPublicKey(Protocol):
    """The key-registry facts required by namespace-aware verification."""

    namespace: str
    public_key: Ed25519PublicKey


@dataclass(frozen=True, slots=True)
class ChainVerificationResult:
    """The completely verified contiguous prefix for one stream."""

    stream_id: str
    start_sequence: int
    end_sequence: int
    last_digest: str
    last_key_id: str
    records: tuple[RecordEnvelope, ...]


def _encode_base64url(value: bytes) -> str:
    return base64.urlsafe_b64encode(value).rstrip(b"=").decode("ascii")


def _decode_base64url(value: object, *, label: str, expected_length: int) -> bytes:
    if (
        not isinstance(value, str)
        or not value
        or re.fullmatch(r"[A-Za-z0-9_-]+", value) is None
    ):
        raise KeyContinuityError(f"{label} must be unpadded base64url")
    try:
        raw = base64.b64decode(
            value + "=" * (-len(value) % 4), altchars=b"-_", validate=True
        )
    except (ValueError, binascii.Error) as exc:
        raise KeyContinuityError(f"{label} must be unpadded base64url") from exc
    if len(raw) != expected_length:
        raise KeyContinuityError(f"{label} has the wrong Ed25519 length")
    return raw


def _public_key_bytes(public_key: Ed25519PublicKey) -> bytes:
    return public_key.public_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PublicFormat.Raw,
    )


def _continuity_payload(
    *,
    predecessor_key_id: str,
    new_key_id: str,
    new_public_key: Ed25519PublicKey,
    tenant_id: str,
    stream_id: str,
) -> dict[str, str]:
    if not all((predecessor_key_id, new_key_id, tenant_id, stream_id)):
        raise KeyContinuityError("continuity IDs must be non-empty strings")
    if predecessor_key_id == new_key_id:
        raise KeyContinuityError("key continuity must describe a real key rotation")
    return {
        "alg": ALGORITHM,
        "predecessor_key_id": predecessor_key_id,
        "new_key_id": new_key_id,
        "new_public_key": _encode_base64url(_public_key_bytes(new_public_key)),
        "tenant_id": tenant_id,
        "stream_id": stream_id,
    }


def create_key_continuity(
    *,
    predecessor_key_id: str,
    predecessor_private_key: Ed25519PrivateKey,
    new_key_id: str,
    new_public_key: Ed25519PublicKey,
    tenant_id: str,
    stream_id: str,
) -> dict[str, str]:
    """Sign a context-bound key rotation with the predecessor key (ES-024a)."""

    payload = _continuity_payload(
        predecessor_key_id=predecessor_key_id,
        new_key_id=new_key_id,
        new_public_key=new_public_key,
        tenant_id=tenant_id,
        stream_id=stream_id,
    )
    return {
        **payload,
        "sig": _encode_base64url(predecessor_private_key.sign(canonicalize(payload))),
    }


def verify_key_continuity(
    assertion: Mapping[str, object],
    *,
    expected_predecessor_key_id: str,
    predecessor_public_key: Ed25519PublicKey,
    expected_new_key_id: str,
    new_public_key: Ed25519PublicKey,
    expected_tenant_id: str,
    expected_stream_id: str,
) -> None:
    """Verify a closed, non-negotiable Ed25519 continuity assertion."""

    if set(assertion) != CONTINUITY_FIELDS:
        raise KeyContinuityError("key continuity assertion has unknown or missing fields")
    if assertion.get("alg") != ALGORITHM:
        raise UnsupportedAlgorithmError(
            "unsupported key-continuity algorithm; v0.1 accepts only ed25519"
        )
    if assertion.get("predecessor_key_id") != expected_predecessor_key_id:
        raise KeyContinuityError("key continuity names the wrong predecessor key")
    if assertion.get("new_key_id") != expected_new_key_id:
        raise KeyContinuityError("key continuity names the wrong new key")
    if assertion.get("tenant_id") != expected_tenant_id:
        raise KeyContinuityError("key continuity tenant_id does not match record tenant_id")
    if assertion.get("stream_id") != expected_stream_id:
        raise KeyContinuityError("key continuity stream_id does not match record stream_id")
    expected_public_key = _public_key_bytes(new_public_key)
    asserted_public_key = _decode_base64url(
        assertion.get("new_public_key"), label="new public key", expected_length=32
    )
    if asserted_public_key != expected_public_key:
        raise KeyContinuityError("key continuity contains the wrong new public key")

    payload = {
        "alg": assertion["alg"],
        "predecessor_key_id": assertion["predecessor_key_id"],
        "new_key_id": assertion["new_key_id"],
        "new_public_key": assertion["new_public_key"],
        "tenant_id": assertion["tenant_id"],
        "stream_id": assertion["stream_id"],
    }
    signature = _decode_base64url(
        assertion.get("sig"), label="continuity sig", expected_length=64
    )
    try:
        predecessor_public_key.verify(signature, canonicalize(payload))
    except InvalidSignature as exc:
        raise KeyContinuityError("new public key is not signed by the predecessor key") from exc


def _gap_body(
    previous: RecordEnvelope,
    current: RecordEnvelope,
    *,
    cause: str,
) -> dict[str, object]:
    return {
        "gap_start": previous.clocks.source_time,
        "gap_end": current.clocks.source_time,
        "affected_scope": {"stream_id": current.stream_id},
        "cause": cause,
        "detection_source": "chain_verifier",
        "exposure": "unknown",
        "actions_during_gap": None,
    }


def _signature_key_id(
    record: RecordEnvelope,
    public_keys: Mapping[str, Ed25519PublicKey],
    *,
    verification_keys: Mapping[str, RegisteredPublicKey],
    valid_through_sequence: int | None = None,
    coverage_gap_body: Mapping[str, object] | None = None,
) -> str:
    try:
        key_id = verify_record_signature(record, public_keys=public_keys)
    except SignatureError as exc:
        raise ChainVerificationError(
            f"signature verification failed at sequence {record.sequence}: {exc}",
            stream_id=record.stream_id,
            break_sequence=record.sequence,
            valid_through_sequence=valid_through_sequence,
            coverage_gap_body=coverage_gap_body,
        ) from exc
    expected = primary_signer_namespace(record.record_type)
    if verification_keys[key_id].namespace != expected:
        raise KeyNamespaceError(
            "key namespace mismatch: "
            f"{record.record_type} requires an {expected} primary signer",
            stream_id=record.stream_id,
            break_sequence=record.sequence,
            valid_through_sequence=valid_through_sequence,
            coverage_gap_body=coverage_gap_body,
        )
    return key_id


def _fork_checked_records(
    records: Iterable[RecordEnvelope],
    public_keys: Mapping[str, Ed25519PublicKey],
    *,
    verification_keys: Mapping[str, RegisteredPublicKey],
) -> list[RecordEnvelope]:
    positions: dict[tuple[str, int], RecordEnvelope] = {}
    for record in records:
        position = (record.stream_id, record.sequence)
        previous = positions.get(position)
        if previous is None:
            positions[position] = record
            continue
        if (
            previous.record_id != record.record_id
            or canonical_digest(previous) != canonical_digest(record)
        ):
            # A fatal fork is an authenticated equivocation.  Unsigned garbage
            # at an occupied position is only a bad submission; allowing it to
            # poison the stream would turn fork detection into a DoS primitive.
            _signature_key_id(
                previous, public_keys, verification_keys=verification_keys
            )
            _signature_key_id(record, public_keys, verification_keys=verification_keys)
            raise StreamForkError(
                f"stream fork at {record.stream_id} sequence {record.sequence}",
                stream_id=record.stream_id,
                break_sequence=record.sequence,
                fatal=True,
            )
    return list(positions.values())


def _verify_stream(
    records: Iterable[RecordEnvelope],
    *,
    public_keys: Mapping[str, Ed25519PublicKey],
    verification_keys: Mapping[str, RegisteredPublicKey],
) -> ChainVerificationResult:
    materialized = _fork_checked_records(
        records, public_keys, verification_keys=verification_keys
    )
    if not materialized:
        raise ChainVerificationError("cannot verify an empty stream")
    stream_ids = {record.stream_id for record in materialized}
    if len(stream_ids) != 1:
        raise ChainVerificationError(
            "verify_evidence_stream accepts exactly one stream_id"
        )
    ordered = sorted(materialized, key=lambda record: record.sequence)
    stream_id = ordered[0].stream_id
    stream_namespace = primary_signer_namespace(ordered[0].record_type)
    for record in ordered[1:]:
        record_namespace = primary_signer_namespace(record.record_type)
        if record_namespace != stream_namespace:
            raise KeyNamespaceError(
                "stream changes primary signer namespace; ES-033 requires a new "
                "stream rather than cross-role continuity",
                stream_id=stream_id,
                break_sequence=record.sequence,
            )

    first = ordered[0]
    if first.sequence != 1:
        raise SequenceGapError(
            f"stream begins at sequence {first.sequence} without a verified anchor",
            stream_id=stream_id,
            break_sequence=first.sequence,
            coverage_gap_body={
                "gap_start": first.clocks.source_time,
                "gap_end": first.clocks.source_time,
                "affected_scope": {"stream_id": stream_id},
                "cause": "sequence_break",
                "detection_source": "chain_verifier",
                "exposure": "unknown",
                "actions_during_gap": None,
            },
        )
    if first.sequence == 1 and first.prev_digest is not None:
        raise DigestLinkError(
            "first stream record must have null prev_digest",
            stream_id=stream_id,
            break_sequence=1,
        )
    last_key_id = _signature_key_id(
        first, public_keys, verification_keys=verification_keys
    )
    if "key_continuity" in first.signature:
        raise KeyContinuityError(
            "key continuity is forbidden on the first stream record",
            stream_id=stream_id,
            break_sequence=1,
        )
    last_digest = canonical_digest(first)
    previous = first

    for current in ordered[1:]:
        if current.sequence != previous.sequence + 1:
            raise SequenceGapError(
                f"sequence gap before {current.sequence}",
                stream_id=stream_id,
                break_sequence=current.sequence,
                valid_through_sequence=previous.sequence,
                coverage_gap_body=_gap_body(previous, current, cause="sequence_break"),
            )
        if current.prev_digest != last_digest:
            raise DigestLinkError(
                f"prev_digest mismatch at sequence {current.sequence}",
                stream_id=stream_id,
                break_sequence=current.sequence,
                valid_through_sequence=previous.sequence,
                coverage_gap_body=_gap_body(previous, current, cause="sequence_break"),
            )

        current_key_id = _signature_key_id(
            current,
            public_keys,
            verification_keys=verification_keys,
            valid_through_sequence=previous.sequence,
            coverage_gap_body=_gap_body(
                previous, current, cause="key_discontinuity"
            ),
        )
        continuity = current.signature.get("key_continuity")
        if current_key_id != last_key_id:
            if not isinstance(continuity, dict):
                raise KeyContinuityError(
                    f"key rotation without continuity at sequence {current.sequence}",
                    stream_id=stream_id,
                    break_sequence=current.sequence,
                    valid_through_sequence=previous.sequence,
                    coverage_gap_body=_gap_body(
                        previous, current, cause="key_discontinuity"
                    ),
                )
            predecessor_key = public_keys.get(last_key_id)
            new_key = public_keys.get(current_key_id)
            if predecessor_key is None or new_key is None:  # already checked, defensive
                raise KeyContinuityError("key continuity refers to an unknown key")
            try:
                verify_key_continuity(
                    continuity,
                    expected_predecessor_key_id=last_key_id,
                    predecessor_public_key=predecessor_key,
                    expected_new_key_id=current_key_id,
                    new_public_key=new_key,
                    expected_tenant_id=current.tenant_id,
                    expected_stream_id=current.stream_id,
                )
            except (KeyContinuityError, UnsupportedAlgorithmError) as exc:
                raise KeyContinuityError(
                    f"invalid key continuity at sequence {current.sequence}: {exc}",
                    stream_id=stream_id,
                    break_sequence=current.sequence,
                    valid_through_sequence=previous.sequence,
                    coverage_gap_body=_gap_body(
                        previous, current, cause="key_discontinuity"
                    ),
                ) from exc
            last_key_id = current_key_id
        elif continuity is not None:
            raise KeyContinuityError(
                f"key continuity without a key rotation at sequence {current.sequence}",
                stream_id=stream_id,
                break_sequence=current.sequence,
                valid_through_sequence=previous.sequence,
                coverage_gap_body=_gap_body(
                    previous, current, cause="key_discontinuity"
                ),
            )

        previous = current
        last_digest = canonical_digest(current)

    return ChainVerificationResult(
        stream_id=stream_id,
        start_sequence=ordered[0].sequence,
        end_sequence=ordered[-1].sequence,
        last_digest=last_digest,
        last_key_id=last_key_id,
        records=tuple(ordered),
    )


def verify_evidence_stream(
    records: Iterable[RecordEnvelope],
    *,
    verification_keys: Mapping[str, RegisteredPublicKey],
) -> ChainVerificationResult:
    """Verify one record stream and enforce ES-033 on every active signer.

    A fork is fatal for the complete stream.  Other breaks expose the last
    valid sequence so an attestation window can terminate there rather than
    incorporating unverifiable evidence.

    There is deliberately no namespace-free variant of this function. One
    existed and was exported as ``verify_stream``: it took bare Ed25519 keys,
    which carry no custody namespace, so it could not establish SE-003.  Any
    caller that reached for the primitive instead of the wrapper bypassed the
    check, and an issuer key authenticating a customer stream is exactly the
    confusion the two namespaces exist to make impossible. Issuer-observation
    streams use the same entry point and are dispatched from record_type. A
    weakening path that is merely unattractive is still a path. This mirrors the same rule
    in ``verifier-go/internal/evidence/chain.go``.
    """

    public_keys = {
        key_id: registered.public_key
        for key_id, registered in verification_keys.items()
    }
    return _verify_stream(
        records,
        public_keys=public_keys,
        verification_keys=verification_keys,
    )


def verify_evidence_streams(
    records: Iterable[RecordEnvelope],
    *,
    verification_keys: Mapping[str, RegisteredPublicKey],
) -> dict[str, ChainVerificationResult]:
    """Verify each stream independently, without inventing cross-stream order.

    The namespace-free ``verify_streams`` was removed for the reason given on
    :func:`verify_evidence_stream`.
    """

    grouped: dict[str, list[RecordEnvelope]] = {}
    for record in records:
        grouped.setdefault(record.stream_id, []).append(record)
    if not grouped:
        raise ChainVerificationError("cannot verify an empty stream collection")
    return {
        stream_id: verify_evidence_stream(
            stream_records, verification_keys=verification_keys
        )
        for stream_id, stream_records in grouped.items()
    }

File: verifier-go/internal/evidence/chain.go
package evidence

import (
	"crypto/ed25519"
	"fmt"

	"github.com/Align3/evidence-control-plane/verifier-go/internal/jcs"
)

// Chain and continuity error codes.
const (
	CodeChainFork               = "chain.fork"
	CodeChainSequenceGap        = "chain.sequence_gap"
	CodeChainPrevDigestMismatch = "chain.prev_digest_mismatch"
	CodeChainStreamMismatch     = "chain.stream_id_mismatch"

	CodeContinuityMissing        = "continuity.missing"
	CodeContinuityUnknownMember  = "continuity.unknown_member"
	CodeContinuityMissingMember  = "continuity.missing_member"
	CodeContinuityTenantMismatch = "continuity.tenant_mismatch"
	CodeContinuityStreamMismatch = "continuity.stream_mismatch"
	CodeContinuitySigInvalid     = "continuity.signature_invalid"
	CodeContinuityFirstRecord    = "continuity.first_record_forbidden"
	CodeContinuityNoKeyChange    = "continuity.no_key_change"
	CodeContinuityKeyMismatch    = "continuity.key_id_mismatch"
	CodeContinuityKeyConflict    = "continuity.registered_key_mismatch"
	CodeContinuityAlgUnsupported = "continuity.algorithm_unsupported"
)

// StreamError reports where a chain stopped being verifiable. CM-017 makes a
// break terminate the window rather than invalidate it wholesale, so the
// sequence that still verified is part of the result, not a detail.
//
// One failure is reported, not a set. Reporting every determinable failure at a
// break is a real improvement and was drafted here, but it is a behavioural
// change that the specification does not yet require and the corpus does not
// yet record; it lands in EV-28, where the specification, the vectors, Python
// and Go move together.
type StreamError struct {
	Code          string
	Msg           string
	BreakSequence int64
	ValidThrough  int64
}

func (e *StreamError) Error() string {
	return fmt.Sprintf("%s at sequence %d: %s", e.Code, e.BreakSequence, e.Msg)
}

// failure is one determinable defect at a break point.
type failure struct {
	code string
	msg  string
}

// breakAt reports the first determinable failure in chain-traversal order.
func breakAt(seq, validThrough int64, failures ...failure) *StreamError {
	for _, f := range failures {
		if f.code != "" {
			return &StreamError{
				Code: f.code, Msg: f.msg,
				BreakSequence: seq, ValidThrough: validThrough,
			}
		}
	}
	return &StreamError{Code: CodeChainSequenceGap, Msg: "unspecified break",
		BreakSequence: seq, ValidThrough: validThrough}
}

// StreamResult describes a fully verified stream.
type StreamResult struct {
	StartSequence int64
	EndSequence   int64
	LastDigest    string
	LastKeyID     string
}

// continuityMembers is the closed member set of ES-024a.
var continuityMembers = map[string]bool{
	"alg": true, "predecessor_key_id": true, "new_key_id": true,
	"new_public_key": true, "tenant_id": true, "stream_id": true, "sig": true,
}

// trustedStreamKey admits one key to a stream, or says why it cannot.
//
// Every key that ever becomes the active signing key passes through here, which
// is the only reason SE-003 holds for a stream rather than for the records
// someone remembered to check. The previous shape — a loop over the declared
// key ids before verification, plus a look at the final key afterwards — left
// two gaps: a key the keyring does not list was skipped by the first check and
// invisible to the second.
//
// asserted is the public key an ES-024a continuity assertion introduced, or nil
// at the sequence-1 trust anchor.
//
// A rotation onto an unlisted key is refused, and that is the substantive
// decision here. The primitive reading of ES-024a is that the predecessor's
// signature is sufficient to trust the successor — and for the *chain* it is:
// the assertion authenticates the new key. It cannot establish the new key's
// namespace, because a namespace is a custody fact about how a key is held,
// stated by the keyring, and no signature made with one key can confer it on
// another. A verifier that accepted the delegation would be reporting SE-003 as
// enforced while the successor's namespace was simply unknown to it. The corpus
// agrees: every rotation vector lists the rotated key in its keyring.
func trustedStreamKey(keys map[string]RegisteredKey, keyID, namespace string,
	asserted ed25519.PublicKey) (ed25519.PublicKey, error) {
	registered, ok := keys[keyID]
	if !ok {
		if asserted == nil {
			return nil, errf(CodeUnknownSigningKey, "no trusted key for %q", keyID)
		}
		return nil, errf(CodeUnknownSigningKey,
			"the rotation to %q is authenticated, but the keyring does not list that "+
				"key and so states no namespace for it; SE-003 is a property of how a "+
				"key is held and cannot be conferred by the key it replaces", keyID)
	}
	if registered.Namespace != namespace {
		return nil, errf(CodeKeyNamespaceMismatch,
			"key %q is in the %q namespace; this stream requires %q keys (ES-033/SE-003)",
			keyID, registered.Namespace, namespace)
	}
	if asserted != nil && !registered.PublicKey.Equal(asserted) {
		return nil, errf(CodeContinuityKeyConflict,
			"the continuity assertion introduces different key material for %q than the "+
				"keyring holds; one key id naming two keys is unresolvable, not a rotation",
			keyID)
	}
	return registered.PublicKey, nil
}

// VerifyEvidenceStream validates links, sequence continuity, forks, and key
// rotation across one stream, in arrival order, and enforces SE-003 on every
// key that authenticates any record in it.
//
// The trust model is per-stream (ES-024a): sequence 1 anchors to the caller's
// trusted keyring, and every later key must be reached by an authenticated
// rotation. Starting a new stream after a tenant rotation proves nothing about
// the old one, so no continuity is inferred across streams.
//
// There is deliberately no namespace-free variant of this function. One existed
// and was exported; the wrapper that added SE-003 around it could be bypassed
// by any caller in the module that reached for the primitive instead, and an
// issuer key authenticating a customer stream is exactly the confusion the two
// namespaces exist to make impossible. A weakening path that is merely
// unattractive is still a path.
func VerifyEvidenceStream(records []*Record, keys map[string]RegisteredKey) (*StreamResult, error) {
	if len(records) == 0 {
		return nil, &StreamError{Code: CodeChainSequenceGap, Msg: "empty stream"}
	}

	// A stream is the records sharing one stream_id (ES-006). Verifying a
	// mixture would compute links and rotations across chains that were never
	// one chain, and report the result as a single verified stream.
	streamID := records[0].StreamID()
	streamNamespace, err := records[0].PrimarySignerNamespace()
	if err != nil {
		return nil, streamErr(err, records[0].Sequence(), 0)
	}
	for _, r := range records {
		if r.StreamID() != streamID {
			return nil, &StreamError{
				Code: CodeChainStreamMismatch, BreakSequence: r.Sequence(),
				Msg: fmt.Sprintf("stream contains records from %q and %q; "+
					"verify one stream at a time", streamID, r.StreamID()),
			}
		}
		recordNamespace, err := r.PrimarySignerNamespace()
		if err != nil {
			return nil, streamErr(err, r.Sequence(), 0)
		}
		if recordNamespace != streamNamespace {
			return nil, &StreamError{
				Code: CodeKeyNamespaceMismatch, BreakSequence: r.Sequence(),
				Msg: fmt.Sprintf("stream changes primary signer namespace from %q to %q; "+
					"ES-033 requires a new stream rather than cross-role continuity",
					streamNamespace, recordNamespace),
			}
		}
	}

	// ES-003: sequence is monotonic within stream_id, starting at 1, no gaps.
	// Without this a stream handed to the verifier starting at 5 verifies
	// clean, and the four records before it are simply not mentioned -- the
	// silent-omission failure the methodology exists to prevent.
	if first := records[0].Sequence(); first != 1 {
		return nil, &StreamError{
			Code: CodeChainSequenceGap, BreakSequence: first, ValidThrough: 0,
			Msg: fmt.Sprintf("stream begins at sequence %d; a stream anchors at 1 "+
				"and everything before this is unaccounted for", first),
		}
	}

	var (
		start        = records[0].Sequence()
		validThrough int64
		prevComplete string
		activeKeyID  string
		activePub    ed25519.PublicKey
		seenSeq      = map[int64]string{}
	)

	for i, r := range records {
		seq := r.Sequence()

		// ES-006: two records at one stream position with differing record_id
		// are a fatal integrity failure, reported rather than resolved. This is
		// checked before anything else because a fork is not a link problem.
		if prevID, seen := seenSeq[seq]; seen {
			if prevID != r.RecordID() {
				return nil, &StreamError{
					Code: CodeChainFork, BreakSequence: seq, ValidThrough: validThrough,
					Msg: "two records occupy this stream position with differing record_id",
				}
			}
			return nil, &StreamError{
				Code: CodeChainFork, BreakSequence: seq, ValidThrough: validThrough,
				Msg: "record replayed at an occupied stream position",
			}
		}
		if i > 0 && seq != records[i-1].Sequence()+1 {
			// ES-008: a gap is detectable and terminates the window here.
			return nil, &StreamError{
				Code: CodeChainSequenceGap, BreakSequence: seq, ValidThrough: validThrough,
				Msg: "sequence is not contiguous with the preceding record",
			}
		}
		seenSeq[seq] = r.RecordID()

		sig, err := r.Signature()
		if err != nil {
			return nil, streamErr(err, seq, validThrough)
		}
		declaredKeyID, _ := objString(sig, "key_id")
		continuityV, hasContinuity := sig.Get("key_continuity")

		// ES-006a: the link is evaluated first, so it leads the report when
		// several failures coincide. It is evaluated *independently* of the key
		// state rather than short-circuiting, because a broken link and an
		// unauthenticated rotation are distinct integrity failures and a
		// relying party needs to see both.
		var linkFail failure
		prev, present := r.PrevDigest()
		switch {
		case i == 0 && present:
			linkFail = failure{CodeChainPrevDigestMismatch,
				"first record of a stream must carry prev_digest null"}
		case i > 0 && !present:
			linkFail = failure{CodeChainPrevDigestMismatch,
				"prev_digest is null but this is not the first record"}
		case i > 0 && prev != prevComplete:
			linkFail = failure{CodeChainPrevDigestMismatch,
				fmt.Sprintf("prev_digest %s does not match the previous complete record %s",
					prev, prevComplete)}
		}

		if i == 0 {
			// Trust anchor. ES-024a forbids a continuity assertion here: there
			// is no predecessor in this stream for it to chain from.
			if hasContinuity {
				return nil, breakAt(seq, validThrough, linkFail, failure{CodeContinuityFirstRecord,
					"sequence 1 anchors to the trusted keyring and must not carry key_continuity"})
			}
			pub, err := trustedStreamKey(keys, declaredKeyID, streamNamespace, nil)
			if err != nil {
				return nil, breakAt(seq, validThrough, linkFail, failure{Code(err), err.Error()})
			}
			activeKeyID, activePub = declaredKeyID, pub
		} else if declaredKeyID != activeKeyID {
			// ES-024: rotation without continuity is a chain break, not a warning.
			if !hasContinuity {
				return nil, breakAt(seq, validThrough, linkFail, failure{CodeContinuityMissing,
					"signing key changed without a key_continuity assertion"})
			}
			newPub, err := verifyContinuity(continuityV, r, activeKeyID, activePub, declaredKeyID)
			if err != nil {
				return nil, breakAt(seq, validThrough, linkFail, failure{Code(err), err.Error()})
			}
			// The assertion authenticates the successor; the keyring is what
			// says it may play the evidence role at all (SE-003).
			pub, err := trustedStreamKey(keys, declaredKeyID, streamNamespace, newPub)
			if err != nil {
				return nil, breakAt(seq, validThrough, linkFail, failure{Code(err), err.Error()})
			}
			activeKeyID, activePub = declaredKeyID, pub
		} else if hasContinuity {
			// ES-024a: an assertion on a record that is not the first after a
			// real key change is rejected, so a replayed proof cannot be parked
			// on an arbitrary record.
			return nil, breakAt(seq, validThrough, linkFail, failure{CodeContinuityNoKeyChange,
				"key_continuity present but the signing key did not change"})
		}

		// A link failure with no accompanying key failure still stops here.
		if linkFail.code != "" {
			return nil, breakAt(seq, validThrough, linkFail)
		}

		if _, err := VerifyRecordSignature(r, map[string]ed25519.PublicKey{activeKeyID: activePub}); err != nil {
			return nil, streamErr(err, seq, validThrough)
		}

		complete, err := r.CompleteDigest()
		if err != nil {
			return nil, streamErr(err, seq, validThrough)
		}
		prevComplete = complete
		validThrough = seq
	}

	return &StreamResult{
		StartSequence: start,
		EndSequence:   validThrough,
		LastDigest:    prevComplete,
		LastKeyID:     activeKeyID,
	}, nil
}

func streamErr(err error, seq, validThrough int64) *StreamError {
	return &StreamError{
		Code: Code(err), Msg: err.Error(),
		BreakSequence: seq, ValidThrough: validThrough,
	}
}

// verifyContinuity checks the ES-024a assertion carried on the first record
// signed by a new key, and returns the newly trusted public key.
//
// The assertion sits inside signature, so the carrying record's own
// signed_digest does not cover it. What authenticates it is the predecessor's
// signature over the other six members; what commits it is the next record's
// prev_digest under ES-006a.
func verifyContinuity(v jcs.Value, r *Record, predecessorKeyID string,
	predecessorPub ed25519.PublicKey, declaredKeyID string) (ed25519.PublicKey, error) {
	cont, ok := v.(*jcs.Object)
	if !ok {
		return nil, errf(jcs.CodeMalformed, "key_continuity must be an object")
	}
	for _, k := range cont.Keys() {
		if !continuityMembers[k] {
			return nil, errf(CodeContinuityUnknownMember,
				"unknown key_continuity member %q (ES-024a)", k)
		}
	}
	for k := range continuityMembers {
		if !cont.Has(k) {
			return nil, errf(CodeContinuityMissingMember,
				"missing key_continuity member %q (ES-024a)", k)
		}
	}
	if alg, _ := objString(cont, "alg"); alg != algorithm {
		return nil, errf(CodeContinuityAlgUnsupported,
			"key_continuity alg must be %q (ES-022)", algorithm)
	}
	// Context binding. Without these two checks a valid proof from one tenant
	// or stream could be replayed into another.
	if got, _ := objString(cont, "tenant_id"); got != r.TenantID() {
		return nil, errf(CodeContinuityTenantMismatch,
			"continuity tenant_id %q does not match the carrying record %q", got, r.TenantID())
	}
	if got, _ := objString(cont, "stream_id"); got != r.StreamID() {
		return nil, errf(CodeContinuityStreamMismatch,
			"continuity stream_id %q does not match the carrying record %q", got, r.StreamID())
	}
	if got, _ := objString(cont, "new_key_id"); got != declaredKeyID {
		return nil, errf(CodeContinuityKeyMismatch,
			"continuity new_key_id %q does not match the signing key %q", got, declaredKeyID)
	}
	if got, _ := objString(cont, "predecessor_key_id"); got != predecessorKeyID {
		return nil, errf(CodeContinuityKeyMismatch,
			"continuity predecessor_key_id %q is not the outgoing key %q", got, predecessorKeyID)
	}
	newPubB64, _ := objString(cont, "new_public_key")
	newPubRaw, err := decodeUnpadded(newPubB64, ed25519.PublicKeySize)
	if err != nil {
		return nil, err
	}
	sigB64, _ := objString(cont, "sig")
	sigRaw, err := decodeUnpadded(sigB64, ed25519.SignatureSize)
	if err != nil {
		return nil, err
	}
	signed, err := jcs.Serialize(cont.Without("sig"))
	if err != nil {
		return nil, err
	}
	if !ed25519.Verify(predecessorPub, signed, sigRaw) {
		return nil, errf(CodeContinuitySigInvalid,
			"predecessor key %q did not sign this continuity assertion", predecessorKeyID)
	}
	return ed25519.PublicKey(newPubRaw), nil
}
